File: Algorithms/15/15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        dp_1 = []
        dp_2 = {}
        ans = []
        if nums.count(0) > 2:
            ans.append([0,0,0])
        for num in nums:
            if not num in dp_1:
                if -num in dp_2.keys():
                    for i,j in dp_2[-num]:
                        ans.append([i,j,num])
                for i in dp_1:
                    dp_2.setdefault(i + num, [])
                    if not sorted([i,num]) in dp_2[i + num]:
                        dp_2[i+num].append(sorted([i,num]))
                dp_1.append(num)
            elif num != 0 :
                dp_2.setdefault(2 * num, [])
                dp_2[2*num].append([num,num])
                if -2*num in dp_1:
                    ans.append([num,num,-2*num]) 
        return ans

File: Algorithms/15/15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<int> dp_1;
        set<vector<int>> dp_2;
        vector<vector<int>> ans;
        for(int i = 0; i < nums.size(); i++){
            for(int j = 0; j < dp_2.size();j++){
                if(dp_2[j][0] + dp_2[j][1] + nums[i] == 0){
                    ans.add(new vector(dp_2[j][0],dp_2[j][1],nums[i]))
                }
            }
            for(int j = 0; j < dp_1.size();j++){
                if(dp_1[j] + nums[i] + nums[i] != 0 and dp_1[j] + dp_1[j] + nums[i] != 0){
                    dp_2.insert(new vector(nums[i],dp_1[j]))
                }
            }
            dp_1.insert(nums[i]);
        }
        return ans;
    }
};